les/utils/limiter.go

```diff
@@ -0,0 +1,405 @@
+	}
+	ag.sumFlatWeight += weight - nq.flatWeight
+	nq.flatWeight = weight
+	ag.groupWeight = ag.sumFlatWeight / uint64(len(ag.nodes))
+	if ag.nodeSelect != nil {
+		ag.nodeSelect.Update(nq)
+	}
+}
+
+// remove removes the node queue from the address group. It is the caller's responsibility
+// to remove the address group from the address map if it is empty.
+func (ag *addressGroup) remove(nq *nodeQueue) {
+	if nq.groupIndex == -1 || nq.groupIndex >= len(ag.nodes) || ag.nodes[nq.groupIndex] != nq {
+		panic("removed node queue is not in this address group")
+	}
+
+	l := len(ag.nodes) - 1
+	if nq.groupIndex != l {
+		ag.nodes[nq.groupIndex] = ag.nodes[l]
+		ag.nodes[nq.groupIndex].groupIndex = nq.groupIndex
+	}
+	nq.groupIndex = -1
+	ag.nodes = ag.nodes[:l]
+	ag.sumFlatWeight -= nq.flatWeight
+	if l >= 1 {
+		ag.groupWeight = ag.sumFlatWeight / uint64(l)
+		if l == 1 {
+			ag.nodeSelect = nil
+		} else {
+			ag.nodeSelect.Remove(nq)
+		}
+	} else {
+		ag.groupWeight = 0
+	}
+}
+
+// choose selects one of the node queues belonging to the address group
+func (ag *addressGroup) choose() *nodeQueue {
+	if ag.nodeSelect == nil { // nodes list should never be empty here
+		return ag.nodes[0]
+	}
+	return ag.nodeSelect.Choose().(*nodeQueue)
+}
+
+// NewLimiter creates a new Limiter
+func NewLimiter(totalPriorLimit uint64) *Limiter {
+	l := &Limiter{
+		addressSelect:   NewWeightedRandomSelect(func(item interface{}) uint64 { return item.(*addressGroup).groupWeight }),
+		valueSelect:     NewWeightedRandomSelect(func(item interface{}) uint64 { return item.(*nodeQueue).valueWeight }),
+		nodes:           make(map[enode.ID]*nodeQueue),
+		addresses:       make(map[string]*addressGroup),
+		totalPriorLimit: totalPriorLimit,
+	}
+	l.cond = sync.NewCond(&l.lock)
+	go l.processLoop()
+	return l
+}
+
+// selectionWeights calculates the selection weights of a node for both the address and
+// the value selector. The weight depends on the relative cost of the previous processed
+// request.
+func selectionWeights(relCost, value float64) (flatWeight, valueWeight uint64) {
+	var f float64
+	if relCost <= 0.001 {
+		f = 1
+	} else {
+		f = 0.001 / relCost
+	}
+	f *= 1000000000
+	flatWeight, valueWeight = uint64(f), uint64(f*value)
+	if flatWeight == 0 {
+		flatWeight = 1
+	}
+	return
+}
+
+// Add adds a new request to the node queue belonging to the given id. Value belongs
+// to the requesting node. A higher value gives the request a higher chance of being
+// served quickly in case of heavy load or a DDoS attack. PriorWeight is a rough estimate
+// of the serving cost of the request. A lower priorWeight also gives the request a
+// better chance.
+func (l *Limiter) Add(id enode.ID, address string, value float64, priorWeight uint64) chan chan float64 {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	process := make(chan chan float64, 1)
+	if l.quit {
+		close(process)
+		return process
+	}
+	if value > l.maxValue {
+		l.maxValue = value
```

Comment:
It looks a bit hacky to me. It means the base(maxValue) for all other requests might be different. I can understand in order to normalize it we have to adjust the cap. 

Why not just passing a "value" in range [0, 1] from the outside? It likes more like a library.

Reply:
The value can be an arbitrarily high number and it is a lot easier to use if we don't expect the caller to know the highest possible value in advance. Changing maxValue while running is not really a problem because it only screws up probability ratios for a short time, each node will be calculated with the right max value next time. I think this solution is convenient and has practically zero risk.